internal/globalconfig/globalconfig.go

```diff
@@ -105,8 +105,6 @@ func SetStatsCarrier(sc *internal.StatsCarrier) {
 
 // PushStat pushes the stat onto the StatsCarrier's stats channel, via the Add method
 func PushStat(stat internal.Stat) {
-	cfg.mu.RLock()
```

Comment:
Isn't this lock needed, though, because `cfg.statsCarrier` can be modified concurrently by `SetStatsCarrier`?

Reply:
> How did you test this? What was failing before?

I ran the package test with this command: `go test -timeout 30s -v gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql/...`.

The tests panicked like this before applying the change:

```
goroutine 738 [sync.RWMutex.Lock]:
sync.runtime_SemacquireRWMutex(0x12addca40?, 0x18?, 0x18?)
	/Users/dario.castane/sdk/go1.21.5/src/runtime/sema.go:87 +0x28
sync.(*RWMutex).Lock(0x14000481eb8?)
	/Users/dario.castane/sdk/go1.21.5/src/sync/rwmutex.go:152 +0xf8
gopkg.in/DataDog/dd-trace-go.v1/internal/globalconfig.SetServiceName({0x1050f2c11, 0xa})
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/internal/globalconfig/globalconfig.go:60 +0x3c
gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/namingschematest.withDDService({0x1050f2c11, 0xa})
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/contrib/internal/namingschematest/namingschematest.go:131 +0x80
gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql.TestNamingSchema.func4.NewServiceNameTest.func3.1(0x0?)
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/contrib/internal/namingschematest/namingschematest.go:88 +0x4c
testing.tRunner(0x14000c501a0, 0x14001db8ac8)
	/Users/dario.castane/sdk/go1.21.5/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 702
	/Users/dario.castane/sdk/go1.21.5/src/testing/testing.go:1648 +0x33c

[...]

goroutine 566 [chan send]:
gopkg.in/DataDog/dd-trace-go.v1/internal.(*StatsCarrier).Add(...)
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/internal/statsd.go:219
gopkg.in/DataDog/dd-trace-go.v1/internal/globalconfig.PushStat({0x10543fea8?, 0x14001a0a7c0?})
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/internal/globalconfig/globalconfig.go:114 +0xe0
gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql.pollDBStats(0x140000af040, {0x14000ae1140, 0x1, 0x1})
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/contrib/database/sql/metrics.go:44 +0x194
created by gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql.OpenDB in goroutine 564
	/Users/dario.castane/go/src/github.com/DataDog/dd-trace-go/contrib/database/sql/sql.go:215 +0x2dc
```

The thing is that the second goroutine (ID 566) locked on sending a payload through a unbuffered channel. It's not an issue for these tests because we don't care about the contrib stats, and there is only one go routine using that channel.

As the goroutine 566 locks on the channel, while holding the `globalconfig` mutext, it causes a deadlock when goroutine 738 tries to lock on the same mutex.

> Isn't this lock needed, though, because cfg.statsCarrier can be modified concurrently by SetStatsCarrier?

Yeah, you are right. I just considered the channel, but cfg.statsCarrier needs still to be protected. Then, the fix must be done in the `cfg.statsCarrier.Add` function.

Reply:
Ah, I see. Perhaps the fix would be to have `StatsCarrier` return `cfg.statsCarrier` directly? Then the caller can check if it's nil and use it without `cfg.mu` locked.

Reply:
@nsrip-dd I see what you mean. I'm going to remove `StatsCarrier` altogether. It's not used anywhere and it adds complexity.

Reply:
I was going to revert my changes in `StatsCarrier.Add`, but I feel more comfortable leaving it as it is. If there is any contention, it'll log it instead of blocking there.

@nsrip-dd @dianashevchenko Can you review it again? Thanks!